### PR TITLE
Add tests with absolute minimum code change

### PR DESCRIPTION
**Pull request by tyrannosaurus-becks**
Hello! Thanks for creating this repository and making it public! 

In light of https://github.com/actions/cache/issues/810 being open long-term, we'd like to move over to using this action. I am a Go developer and would like to harden this repo's code first. I thought I'd start by adding test coverage for the code's existing state, and after that exists, I might follow up with a second PR further hardening the code. 

If that seems cool with you, it would be awesome if you'd consider this PR. Please let me know if there's anything you'd like me to change to make it better, I'd be happy to do that.

Thanks in advance! 😄 ❤️

### CODE REVIEW

File: s3.go [hunk 1]
```diff
@@ -12,11 +12,13 @@ import (
 	"github.com/pkg/errors"
 )
 
+var session = func() S3Session {
+	cfg, _ := config.LoadDefaultConfig(context.TODO())
+	return s3.NewFromConfig(cfg)
+}()
```
Comment (tyrannosaurus-becks) [inline s3.go:18]: In this PR, I move session creation _out_ of the methods themselves and make the session into an interface. That way, I can swap out the interface with a mock to add tests that don't really hit S3.

File: s3.go [hunk 2]
```diff
@@ -40,32 +42,30 @@ func PutObject(key, bucket, s3Class string) error {
...
-	session := s3.NewFromConfig(cfg)
-
 	i := &s3.GetObjectInput{
 		Bucket: aws.String(bucket),
 		Key:    aws.String(key),
 	}
 
-	size, err := session.GetObject(context.TODO(), i)
-
-	log.Printf("Cache downloaded successfully, containing %d bytes", size)
```
Comment (tyrannosaurus-becks) [inline s3.go:53]: My tests initially failed because it turned out that `session.GetObject` doesn't return a `size` int, it actually returns a full-fledged object. There's no direct way to get the size of the object from what's returned, so I tried to do the closest thing possible - displaying the Content Range if it exists.

File: types.go [hunk 3]
```diff
@@ -1,5 +1,18 @@
...
+
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+)
+
+type S3Session interface {
+	PutObject(ctx context.Context, params *s3.PutObjectInput, optFns ...func(*s3.Options)) (*s3.PutObjectOutput, error)
+	GetObject(ctx context.Context, params *s3.GetObjectInput, optFns ...func(*s3.Options)) (*s3.GetObjectOutput, error)
+	DeleteObject(ctx context.Context, params *s3.DeleteObjectInput, optFns ...func(*s3.Options)) (*s3.DeleteObjectOutput, error)
+	HeadObject(ctx context.Context, params *s3.HeadObjectInput, optFns ...func(*s3.Options)) (*s3.HeadObjectOutput, error)
+}
```
Comment (tyrannosaurus-becks) [inline types.go:14]: This is the interface being used as the S3 session so we can either use a _real_ session, or a mocked one for unit tests.

**Comment by tyrannosaurus-becks** [comment]
@gmsantos @amendoa @brunopadz if any of you are available to check this out it would be much appreciated.

File: s3_test.go [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+	}
+	return &s3.DeleteObjectOutput{}, nil
+}
+
+func (m *mockedS3Session) HeadObject(ctx context.Context, params *s3.HeadObjectInput, optFns ...func(*s3.Options)) (*s3.HeadObjectOutput, error) {
+	if ctx == nil {
+		return nil, errors.New("context is nil")
+	}
+	if *params.Bucket != m.ExpectedBucket {
+		// TODO is this correct? There is also a types.NoSuchBucket in the AWS SDK.
```
Comment (tyrannosaurus-becks) [inline s3_test.go:258]: It might be a bug that we're not also looking for a `NoSuchBucket` error in the `ObjectExists` method, but since I'm trying to do the absolute minimum code change, I'm leaving out testing for it and potentially fixing it, because that would be in a separate PR.

Comment (tyrannosaurus-becks) [reply s3_test.go:258]: Follow-up, I did find that this isn't what AWS returns when a bucket or file is missing. After this PR is accepted and merged, I have an immediate follow-on PR I can post that adds _even more_ tests and fixes this! 😄

**Comment by gmsantos** [comment]
@tyrannosaurus-becks sorry I'm not a maintainer. At the moment, I think @amendoa is the only maintainer.

**Comment by brunopadz** [comment]
Same here @tyrannosaurus-becks, not a maintainer anymore.

**Comment by tyrannosaurus-becks** [comment]
Closing this since this repo doesn't appear to be maintained.
